autoscheduler/frontend/src/redux/reducers.ts

```diff
@@ -61,10 +65,120 @@ function courseCards(
+// in the future, this could also support preferred times
+function availabilityMode(
+  state: AvailabilityType = AvailabilityType.NONE, action: AvailabilityModeAction,
+): AvailabilityType {
+  if (action.type === SET_AVAILABILITY_MODE) { return action.mode; }
+  return state;
+}
+
+// stores all information about the user's availability
+function availability(
+  state: Availability[] = [], action: AvailabilityAction,
+): Availability[] {
+  switch (action.type) {
+    case ADD_AVAILABILITY:
+    case MERGE_AVAILABILITY: {
+      // helper functions to get the start and end times of an availability object in minutes
+      const getStart = (av: Availability): number => av.startTimeHours * 60 + av.startTimeMinutes;
+      const getEnd = (av: Availability): number => av.endTimeHours * 60 + av.endTimeMinutes;
+
+      // check for overlaps between the availability to be added and pre-existing ones
+      let overlapsFound = 0; // counts merging overlaps only
+      let overlapIdx = -1;
+
+      const initNewAv = action.type === ADD_AVAILABILITY
+        ? argsToAvailability(action.availability)
+        : state[state.length - 1];
+      let newAv = initNewAv;
+      return state.reduce<Availability[]>((avsList, oldAv, idx): Availability[] => {
+        // only counts as an overlap if they're on the same day of the week
+        if (oldAv.dayOfWeek === newAv.dayOfWeek) {
+          const avWithEarlierStart = getStart(oldAv) < getStart(newAv) ? oldAv : newAv;
+          const avWithLaterStart = avWithEarlierStart === oldAv ? newAv : oldAv;
+          // overlap detected
+          if (getEnd(avWithEarlierStart) >= getStart(avWithLaterStart)) {
+            // if they're the same type, merge them
+            if (oldAv.available === newAv.available) {
+              overlapsFound += 1;
+              const oldOverlapIdx = overlapIdx;
+              overlapIdx = idx;
+              const newEnd = Math.max(getEnd(oldAv), getEnd(newAv));
+              const newEndHrs = Math.floor(newEnd / 60);
+              const newEndMins = newEnd % 60;
+              newAv = {
+                available: oldAv.available,
+                dayOfWeek: oldAv.dayOfWeek,
+                startTimeHours: avWithEarlierStart.startTimeHours,
+                startTimeMinutes: avWithEarlierStart.startTimeMinutes,
+                endTimeHours: newEndHrs,
+                endTimeMinutes: newEndMins,
+              };
+              if (overlapsFound === 2) {
+                return avsList.slice(0, oldOverlapIdx)
+                  .concat(avsList.slice(oldOverlapIdx + 1))
+                  .concat(newAv);
+              }
+              return avsList.filter((av) => av !== oldAv).concat(newAv);
+            }
+            // if they're different types, then set the new one to the old one's borders
+            // TODO
+          }
+        }
+        // if no overlap
+        return avsList.concat(oldAv);
+      }, []).concat(overlapsFound ? [] : initNewAv);
```

Comment:
Unfortunately, that would be bad practice with Redux, since `.push()` would mutate the argument array. I'm not sure why mutations are so bad, but the Redux guide recommended avoiding them, which is why I have the `concat`s

Reply:
If I'm not mistaken, redux only wants you to avoid mutating any part of the state (so that time travel is possible) so since `avsList` is mutating the accumulator of the reduce function rather than part of the state, it should be find to use `push()`, since calling this with the same state/action will always result in the same value

Reply:
そうです。ちがったので、ごめんなさい。

Reply:
謝っても、許せない。アデルさんはこのプロジェクトの不名誉だ。

冗談だけｗｗ